cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,662 @@
+  // Examples:
+  //     modulo(1,60)    1
+  //     modulo(-1,60)  59
+  auto modulo_time = [](int64_t time, int64_t base) {
+    return static_cast<int32_t>(((time % base) + base) % base);
+  };
+
+  // This function handles converting units by dividing and adjusting for negative values.
+  // Examples:
+  //     scale(-61,60)  -2
+  //     scale(-60,60)  -1
+  //     scale(-59,60)  -1
+  //     scale( 59,60)   0
+  //     scale( 60,60)   1
+  //     scale( 61,60)   1
+  auto scale_time = [](int64_t time, int64_t base) {
+    return static_cast<int32_t>((time - ((time < 0) * (base - 1L))) / base);
+  };
+
+  int64_t seconds{0};
+  if (units == type_id::DURATION_SECONDS) {
+    seconds = duration;
+  } else if (units == type_id::DURATION_MILLISECONDS) {
+    seconds = simt::std::chrono::floor<cudf::duration_s>(cudf::duration_ms(duration)).count();
+    timeparts[DU_SUBSECOND] = modulo_time(duration, 1000L);
+  } else if (units == type_id::DURATION_MICROSECONDS) {
+    seconds = simt::std::chrono::floor<cudf::duration_s>(cudf::duration_us(duration)).count();
+    timeparts[DU_SUBSECOND] = modulo_time(duration, 1000L * 1000);
+  } else if (units == type_id::DURATION_NANOSECONDS) {
+    seconds = simt::std::chrono::floor<cudf::duration_s>(cudf::duration_ns(duration)).count();
+    timeparts[DU_SUBSECOND] = modulo_time(duration, 1000L * 1000 * 1000);
+  }
+  timeparts[DU_DAY]    = scale_time(seconds, 24 * 60 * 60);
+  timeparts[DU_HOUR]   = modulo_time(scale_time(seconds, 60 * 60), 24);
+  timeparts[DU_MINUTE] = modulo_time(scale_time(seconds, 60), 60);
+  timeparts[DU_SECOND] = modulo_time(seconds, 60);
+}
+
+template <typename T>
+struct duration_to_string_size_fn {
+  const column_device_view d_durations;
+  const format_item* d_format_items;
+  size_type items_count;
+  type_id type;
+
+  __device__ int8_t count_trailing_zeros(int n) const
+  {
+    int8_t zeros = 0;
+    if ((n % 100000000) == 0) {
+      zeros += 8;
+      n /= 100000000;
+    }
+    if ((n % 10000) == 0) {
+      zeros += 4;
+      n /= 10000;
+    }
+    if ((n % 100) == 0) {
+      zeros += 2;
+      n /= 100;
+    }
+    if ((n % 10) == 0) { zeros++; }
+    return zeros;
+  }
+
+  __device__ int8_t format_length(char format_char, int32_t const* const timeparts) const
+  {
+    switch (format_char) {
+      case 'd': return count_digits(timeparts[DU_DAY]); break;
+      case '+': return timeparts[DU_DAY] < 0 ? 1 : 0; break;
+      case 'H': return count_digits(timeparts[DU_HOUR]); break;
+      case 'M': return count_digits(timeparts[DU_MINUTE]); break;
+      case 'S': return count_digits(timeparts[DU_SECOND]); break;
+      // 0 or 6 digits for pandas, include dot only if non-zero.
+      case 'u': return (timeparts[DU_SUBSECOND] == 0) ? 0 : 6 + 1; break;
+      // 0 or ns without trailing zeros
+      case 'f':
+        return (timeparts[DU_SUBSECOND] == 0) ? 0 : [units = type] {
+          if (units == type_id::DURATION_MILLISECONDS) return 3 + 1;
```

Comment:
The `+1` here are for including the dot?

Reply:
Yes.

Reply:
Please add a comment.